File: C/Partie4/Exo12.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/wait.h>

int main()
{
    pid_t pid;
    int tube1[2], tube2[2];
    pipe(tube1);
    pipe(tube2);
    pid = fork();
    unsigned char buffer[50];
    int result;

    if (pid == 0)
    {
        close(tube1[0]);
        close(tube2[1]);
        printf("Le mot que vous recherchez: ");
        fgets(buffer, 50, stdin);
        buffer[strcspn(buffer, "\n")] = '\0';
        write(tube1[1], buffer, 50);
        read(tube2[0], &result, sizeof(int));
        if (result)
            printf("Le mot a ete trouve\n");
        else
            printf("Le mot n'existe pas\n");
    }
    else if (pid > 0)
    {
        int valeur;
        FILE *fichier = fopen("texte.txt", "r");
        char *word = malloc(50 * sizeof(char));
        close(tube1[1]);
        close(tube2[0]);
        read(tube1[0], buffer, 50);
        while (fscanf(fichier,"%s",word) == 1)
        {
            if (strcmp(word, buffer) == 0)
            {
                valeur = 1;
                break;
            }
            else
                valeur = 0;
        }
        fclose(fichier);
        write(tube2[1], &valeur, sizeof(int));
        wait(NULL);
    }
        return 0;
}

File: Python/Partie4/Exo12.py
import os

tube1 = os.pipe()  # fils->père
tube2 = os.pipe()  # père->fils

pid = os.fork()
if pid == 0:  # fils
    os.close(tube1[0])
    os.close(tube2[1])
    mot = input("Le mot que vous recherchez: ")
    os.write(tube1[1], mot.encode())
    os.close(tube1[1])
    result = int.from_bytes(os.read(tube2[0], 4), 'little')
    if result:
        print("Le mot a été trouvé")
    else:
        print("Le mot n'existe pas")
else:  # père
    os.close(tube1[1])
    os.close(tube2[0])
    mot_recu = os.read(tube1[0], 50).decode().strip('\x00')
    trouve = 0
    try:
        with open("texte.txt", "r") as f:
            for ligne in f:
                for mot in ligne.split():
                    if mot == mot_recu:
                        trouve = 1
                        break
                if trouve:
                    break
    except FileNotFoundError:
        pass
    os.write(tube2[1], trouve.to_bytes(4, 'little'))
    os.close(tube2[1])
    os.wait()